Support Windows Paths

peterpostmann opened the pull request:
Hi,
I tried to use the library with Windows but it dosn't work. I think the main problem is PHP in this case. `__DIR__` should return a trailing slash on Windows as well, but I think that's not possible for legacy reasons. I'm not sure if PHP is just super inconsistent here or if I'm missing something.

Here is what I found out and how I attempted to fix it:

All files in the test cases and examples are loaded using `'file://' . __DIR__ . '/file.json'` which resolvs to something like `file://C:\folder\subfolder/file.json`. This URI is incorrect for 2 reasons:

- There shouldn't be any backslashes
`parse_url` (which is used by sabre-io/uri) returns false if there are backslashes. The backslashes must be converted to normal slashes. Using the provided examples and test cases on windows causes a `Sabre\Uri\InvalidUriException`
- The format is `file://<host>/<path>` but the host is not present if `__DIR__` returns a path on a drive on Windows (if it returns a path from the network the host is present). The (emtpy) host me added (which results in a third slash after `file:`): `file:///C:/folder/subfolder/file.json`. If the third slash is not added, `parse_url` will interprete this as host:port which results in C:empty. The colon is lost and this information can not be recovered from the result of `parse_url` causing `League\JsonReference\SchemaLoadingException` when external reference is loaded (e.g. file a.sjon references b.json: the loader will try to load `file://c/folder/subfolder/b.json`).

Basically the path provided to the function is wrong. Instead of `'file://' . __DIR__ . '/file.json'`, there should be a logic to correctly determine the host part and convert backslashes. Since this format is used in all test cases, the examples and possible also by everyone else using this lib I think it would be a bad idea to burden this on the user of the library.

Hence I suggest to handle this in the library. Therefore `Derferencer.php prepareArguments` should replace blackslashes and add the missing slash for windows path.

This will result in a different behavior: `functions.php parse_external_ref` will return a path with a trailing slash (`/C:/folder/subfolder/file.json`). This cannot be opened using `file_get_contents`. Therefore the loader must be changed as well. 

The library splitts the URI into prefix and path. But the path is not what `parse_url` returns as path, but more: it includes host, port, username, password if present. Therefore the file loader should handle the path the same way, the CurlWebLoader does it: `$uri = $this->prefix . $path;`

This change would change the way the URI is interpreted, hence it may brake something if the loader is used from other interfaces as well. Therefore I suggest to implement the same change for the loader as well, so that it can handle both paths (`C:/folder/subfolder/file.json` and `/C:/folder/subfolder/file.json`).

This is a breaking change but it may not have any consequences because it only affects the behavior on Windows and the library was not working on Windows anyway in first place.

I don't see a good way to test this. Basically many of the test cases are failing right now, if you run them on Windows.

Looking forward to feedback!

peterpostmann committed:
Support windows paths

matt-allan commented:
Hello,

That is unfortunate, I didn't realize windows would return an invalid file URI but I suppose that makes sense.

It seems like the root cause is we aren't passing a valid file URI in to the dereferencer in the first place.  I would rather fix the URI before it's passed in.  It shouldn't be the library's responsibility to fix malformed URIs.

From reading this [Windows article](https://blogs.msdn.microsoft.com/ie/2006/12/06/file-uris-in-windows/) it looks like you are on the right track with the steps necessary to transform the file path into a valid URI.  I think it would make sense to extract that logic to a helper function and call that with the path to get the file URI:

```php
$dereferencer->dereference(file_uri(__DIR__ . '/schema.json'));
```

Reading the Microsoft article it sounds like there are quite a few steps to transform the file path to a valid URI and they don't seem to be fully documented anywhere.  if it ends up being complex I might want to extract that to a separate library.

> This will result in a different behavior: functions.php parse_external_ref will return a path with a trailing slash (/C:/folder/subfolder/file.json). This cannot be opened using file_get_contents. Therefore the loader must be changed as well.

Do you mean the _leading_ slash (before `C:`)?  The Microsoft article seems to indicate the leading slash is supposed to be there, so adding logic to the FileLoader to handle that if the OS is windows seems fine.

> The library splitts the URI into prefix and path. But the path is not what parse_url returns as path, but more: it includes host, port, username, password if present. Therefore the file loader should handle the path the same way, the CurlWebLoader does it: $uri = $this->prefix . $path;

I see how this could be a problem but is it currently contributing to the problem with loading windows file URIs?

I will try and setup Windows CI when I get a chance so we can start testing on Windows.

peterpostmann commented:
Hey,

I extracted the functionality: https://github.com/peterpostmann/php-fileuri

Would be nice if you could have a look and tell me if you like it. If so, I will change the code of this PR.

Regarding your comments:

> Do you mean the leading slash (before C:)? The Microsoft article seems to indicate the leading slash is supposed to be there, so adding logic to the FileLoader to handle that if the OS is windows seems fine.

There should be a leading slash if it is an URI, e.g. `file:///C:/path/file.txt`, but if you call `list($prefix, $path) = explode('://', $ref, 2);` this will result in `file` and `/C:/path/file.txt`. 

> I see how this could be a problem but is it currently contributing to the problem with loading windows file URIs?

If windows URIs are loaded correctly, they cannot be used the way they are used now, because `file_get_contents('/C:/path/file.txt')` fails. The function works with the full URI `file_get_contents('file:///C:/path/file.txt')` or a valid Windows path `file_get_contents('C:/path/file.txt')`.

matt-allan commented:
Hello,

The lib looks good.  I think maybe we could add that to composer's `suggest` and drop a link in the docs for windows users?

It would also be great to use that to get the tests working cross platform.  I am working on setting up CI builds on windows right now.

> The function works with the full URI 

Hm, does that mean we could re-append the `file://` prefix in the `FileLoader` and it would work correctly if they initially passed in the URI returned from your function?

peterpostmann commented:
Hi,

it took some time to work through all the things. Let me share some thought:

I tried to use file:// URIs for everything, but php uses multiple other formats for valid file URIs (http://php.net/manual/en/wrappers.file.php), like
- /path/to/file.ext
- relative/path/to/file.ext
- fileInCwd.ext
- C:/path/to/winfile.ext
- C:\path\to\winfile.ext
- \\smbserver\share\path\to\winfile.ext
- file:///path/to/file.ext
so I concluded it is better to support those URIs (which can also be opened by file_get_contents), than forcing everyone to convert them to file URIs.

Since sabre/uri does not support windows paths, I recreated the library with support for this URIs. The first change therefore was to replace the parse and resolve URI functions. But this is not sufficent:

In case of 
- C:/path/to/winfile.ext
- C:\path\to\winfile.ext
- \\smbserver\share\path\to\winfile.ext
there is no prefix to split from the path to determine what is the correct loader. Therefore the prefix is created in this case. If a prefix is present, it is removed from the path to be compatible to the existing interfaces. 

I removed "file://" from the test cases, since the lib now supports C:\something or /something. Furthermore I added an empty query to certain URIs: $path = __DIR__ . '/fixtures/schema.json?#/properties';

This is needed becuase # is a valid character for windows file names. Therefore, you can either:
- add an empty query
- provide a file URI
But C:\file#abc will be interpreted as "file#abc" file but not "file" + fragment "abc".

I also altered the file loader, since file URIs can have query paramenters (file://abc?xyz) but file_get_contents does not support them.

matt-allan commented:
Hi,

The reason for requiring file Uris instead of any path php can support is because that is what the join reference specification requires.  If you use any file path php can parse instead of a uri your schemas are no longer compatible with other libraries or languages and resolution becomes much more difficult.  For example, the behavior when encountering a relative http uri when the previous schema was loaded from a file path would be undefined, but it works with a file uri. 

The reason I suggested a function to convert the path into a file uri is because the library remains compliant with the spec, uri resolution still works, and it's still easy for the user.

I started getting appveyor setup for Ci builds (had to resolve some GitHub oauth permission issues with work). I am going to close this PR but it would be great to get the tests  passing on Windows by using your library and making any necessary changes to the file loader.

nobbynobbs commented:
Hello again.

May I ask the question?

I have league/json-reference package installed with composer, and it collapse when I pass shema uri like `file:///D:/full/path/to/json/file.json` (throws SchemaLoadingException ( The schema "/D:/full/path/to/json/file.json" could not be found)).

Method \League\JsonReference\Loader\FileLoader::load in vendor folder looks like:
```php
public function load($path)
{
    if (!file_exists($path)) {
        throw SchemaLoadingException::notFound($path);
    }
    return $this->jsonDecoder->decode(file_get_contents($path));
}
```

In the branch master of this repository the same method looks like:
```php
public function load($path)
{
    $uri = 'file://' . $path;
 if (!file_exists($uri)) {
        throw SchemaLoadingException::notFound($uri);
    }
    return $this->jsonDecoder->decode(file_get_contents($uri));
}
```
and it works fine in my case (I replaced it manually in vendor folder). Schema loaded and references successfully resolves. 

Could you explain please, why branch master and published version are not equal?

matt-allan commented:
The fix you are seeing was merged in #17. I have not tagged a new release since that was merged.  I will tag a new release soon.
